homeassistant/helpers/service.py

```diff
@@ -845,3 +848,35 @@ async def execute_service(self, service_call: ServiceCall) -> None:
             async with self._service_condition:
                 self._service_running = False
                 self._service_condition.notify_all()
+
+
+@bind_hass
+@callback
+def async_get_config_entry_from_call(
+    hass: HomeAssistant, call: ServiceCall
+) -> ConfigEntry | None:
+    """Get the config entry related to a particular service call.
+
+    We start narrow and go broad:
+       1. If one or more entity IDs match a config entry, use the first match.
+       2. If one or more device IDs match a config entry, use the first match.
+       3. If one or more area IDs match a config entry, use the first match.
+       4. If no matches, return None.
+    """
+    for module, key in (
+        (entity_registry, CONF_ENTITY_ID),
+        (device_registry, CONF_DEVICE_ID),
+        (area_registry, CONF_AREA_ID),
+    ):
+        if (targets := call.data.get(key)) is None:
+            continue
+
+        registry = module.async_get(hass)
+
+        for target_id in targets:
```

Comment:
I'm assuming that service calls always ensure that entity/device/area IDs are wrapped in a list?